rfcs/0000-graphql-api.md

```diff
@@ -0,0 +1,411 @@
+need to store information about blocks that these keys have created.
+We will likely want to at least optionally cache the history of the blocks/payments on the client side as well,
+just to avoid needing to query it all every time (though we could also paginate the )
+
+Note: There are several custom scalars defined at the top for readability. These will be `String`s in the final implementation due to the complications involved in the encoding of custom scalars not being expressed in the schema and needing to be implemented symmetrically on the client and server.
+
+```graphql
+scalar Date
+scalar PublicKey
+scalar PrivateKey
+scalar Int64
+
+enum ConsensusStatus {
+  SUBMITTED
+  INCLUDED
+  FINALIZED
+  SNARKED
+  FAILED
+}
+
+type ConsensusState {
+  status: ConsensusStatus!
+  estimatedPercentConfirmed: Float!
+}
+
+type Payment {
+  nonce: Int!,
+  submittedAt: Date!
+  includedAt: Date
+  from: PublicKey! 
+  to: PublicKey!
+  amount: Int64!
+  fee: Int!,
+  memo: String
+}
+
+type PaymentUpdate {
+  payment: Payment
+  consensus: ConsensusState!
+}
+
+enum SyncStatus {
+  ERROR
+  BOOTSTRAP # Resyncing
+  CATCHUP # Synced but a little out
+  STALE # You haven't seen any activity recently
+  SYNCED
+}
+
+type SyncUpdate {
+  status: SyncStatus!
+  estimatedPercentSynced: Float!
+  description: String
+}
+
+type SnarkWorker {
+  key: PublicKey!
+  fee: Int!
+}
+
+type SnarkFee {
+  snarkCreator: PublicKey!
+  fee: Int!
+}
+
+type SnarkFeeUpdate {
+  fee: SnarkFee
+  consensus: ConsensusState!
+}
+
+type Block {
+  coinbase: Int!
+  creator: PublicKey!
+  payments: [Payment]!
+  snarkFees: [SnarkFee]!
+}
+
+type BlockUpdate {
+  block: Block!
+  consensus: ConsensusState!
+}
+
+type Wallet {
+  publicKey: PublicKey!
+  privateKey: PrivateKey!
+  balance(consensus: ConsensusStatus): Int64!
+}
+
+type NodeStatus {
+  network: String
+}
+
+## Input types
+
+input AddWalletInput {
+  public: PublicKey
+  private: PrivateKey
+}
+
+input DeleteWalletInput {
+  public: PublicKey
+}
+
+input AddPaymentReceiptInput {
+  receipt: String
+}
+
+input SetNetworkInput {
+  address: String
+}
+
+input SetSnarkWorkerInput {
+  worker: PublicKey!
+  fee: Int!
+}
+
+input CreatePaymentInput {
+  from: PublicKey!,
+  to: PublicKey!,
+  amount: Int64!,
+  fee: Int!,
+  memo: String
+}
+
+input PaymentFilterInput {
+  toOrFrom: PublicKey,
+}
+
+input BlockFilterInput {
+  creator: PublicKey,
+}
+
+## Payload types
+
+type CreatePaymentPayload {
+  payment: Payment
+}
+
+type SetSnarkWorkerPayload {
+  worker: SnarkWorker
+}
+
+type SetNetworkPayload {
+  address: String
+}
+
+type AddPaymentReceiptPayload {
+  payment: Payment
+}
+
+type AddWalletPayload {
+  publicKey: PublicKey
+}
+
+type DeleteWalletPayload {
+  publicKey: PublicKey
+}
+
+# Pagination types
+
+type PageInfo {
+  hasPreviousPage: Boolean!
+  hasNextPage: Boolean!
+}
+
+type PaymentEdge {
+  cursor: String
```

Comment:
I made a note in the RFC about how we should consider avoiding custom scalars in general, but for this specific case, it might be ok since none of the clients will need to handle it directly (especially if it serializes as a string)